src/full/Agda/TypeChecking/MetaVars/Occurs.hs

```diff
@@ -521,30 +521,31 @@ instance Occurs Term where
           Sort s      -> Sort <$> do underRelevance NonStrict $ occurs_ s
           MetaV m' es -> do
             m' <- metaCheck m'
-            -- The arguments of a meta are in a flexible position
-            (MetaV m' <$> do flexibly $ occurs es) `catchError` \ err -> do
-                ctx <- ask
-                reportSDoc "tc.meta.kill" 25 $ vcat
-                  [ text $ "error during flexible occurs check, we are " ++ show (ctx ^. lensFlexRig)
-                  , text $ show err
-                  ]
-                case err of
-                  -- On pattern violations try to remove offending
-                  -- flexible occurrences (if not already in a flexible context)
-                  PatternErr{} | not (isFlexible ctx) -> do
-                    reportSLn "tc.meta.kill" 20 $
-                      "oops, pattern violation for " ++ prettyShow m'
-                    -- Andreas, 2014-03-02, see issue 1070:
-                    -- Do not prune when meta is projected!
-                    caseMaybe (allApplyElims es) (throwError err) $ \ vs -> do
-                      killResult <- lift . prune m' vs =<< allowedVars
-                      if (killResult == PrunedEverything) then do
-                        -- after successful pruning, restart occurs check
-                        reportSDoc "tc.meta.prune" 40 $ "Pruned everything"
-                        v' <- instantiate (MetaV m' es)
-                        occurs v'
-                      else throwError err
-                  _ -> throwError err
+            ctx <- ask
+            let fallback = return . Left $ MetaV m' es
+            -- Andreas, 2014-03-02, see issue 1070:
+            -- Do not prune when meta is projected!
+            t <- caseMaybe (allApplyElims es) (fallback) $ \ vs -> do
+              -- don't attempt pruning if we're in a flexible context
+               if not (isFlexible ctx) then do
+                 killResult <- lift . prune m' vs =<< allowedVars
+                 v <- instantiate (MetaV m' es)
+                 if (killResult == PrunedEverything) then do
+                   reportSDoc "tc.meta.prune" 40 $ "Pruned everything"
+                   -- we don't have to do occurs-check after everything was pruned
+                   return $ Right v
+                 else do
+                   reportSDoc "tc.meta.prune" 40 $ "Didn't manage to prune everything"
+                   -- some variables was pruned, but not all, still have to do occurs-check after
+                   return $ Left v
+               else fallback
+            case t of
+              -- The arguments of a meta are in a flexible position
+              (Left (MetaV m' es')) -> (MetaV m' <$> do flexibly $ occurs es')
+              -- Sometimes instantiated meta is actually not a meta itself
+              (Left t) -> occurs t
+              -- the PrunedEverything case
+              (Right v) -> return v
```

Comment:
The old occurs check would restart on the pruned `v`, the new one does not.  Is it clear that the result of `PrunedEveryting` is always passing the occurs check?  If yes, it should be documented as postcondition in the `prune` function.